src/inewsFtp.ts

```diff
@@ -95,264 +88,161 @@ export class INewsClient extends EventEmitter {
 		})
 	}
 
-	connect(callback: NodeCallback<FtpClient>, forceDisconnect?: boolean): void {
-		let reconnectAttempts = 0
+	async connect(): Promise<FtpClient> {
+		// If there's a connection in-progress, return that promise.
+		if (this._connectionPromise) {
+			return this._connectionPromise
+		}
 
-		const attemptReconnect = () => {
-			if (forceDisconnect || reconnectAttempts > 0) {
-				this.disconnect(() => {
-					connect(connectResult)
-				})
-			} else connect(connectResult)
+		// If we're already connected, return that connection.
+		if (this._ftpConn !== null && this._ftpConn.connected) {
+			return this._ftpConn
 		}
 
-		const connect = (connectResult: NodeCallback<FtpClient>) => {
-			let returned = false
+		// Else, connect.
+		this._currentDir = null
+		this._connectionPromise = new Promise((resolve, reject) => {
+			let reconnectAttempts = 0
 
-			const onReady = () => {
-				if (!returned) {
-					returned = true
-					this._currentDir = null
-					removeListeners()
-					connectResult(null, this._ftpConn)
-				}
-			}
+			const attemptReconnect = async (): Promise<FtpClient> => {
+				let promise: Promise<FtpClient>
 
-			const onError = (error: NodeJS.ErrnoException) => {
-				if (!returned) {
-					returned = true
-					removeListeners()
-					connectResult(error)
+				if (reconnectAttempts > 0) {
+					promise = this.disconnect().then(async () => {
+						return connect()
+					})
+				} else {
+					promise = connect()
 				}
-			}
 
-			const onEnd = () => {
-				if (!returned) {
-					returned = true
-					removeListeners()
-					connectResult(null, this._ftpConn)
+				reconnectAttempts++
+
+				try {
+					const ftpConn = await promise
+					return ftpConn
```

Comment:
This can be simplified to
```suggestion
					return promise
```